R/adjust-mh.R

```diff
@@ -0,0 +1,173 @@
+  var.est.GR = sum(var.k[weight!=0])/(sum(weight)^2)
+  return(var.est.GR)
+}
+
+
+
+var.mh.mGR <- function(n11k, n10k, n.1k, n.0k){
+  help_f = Vectorize(function(x){
+    if(x>1){return( x/(x-1))}else {1}
+  })
+  
+  n01k = n.1k - n11k
+  n00k = n.0k - n10k
+  weight <- n.1k * n.0k/ (n.1k + n.0k)
+  var.k = weight^2 * (n11k * n01k / n.1k^3 * help_f(n.1k) + n10k*n00k/n.0k^3*help_f(n.0k))
+  var.est.mGR = sum(var.k[weight!=0])/(sum(weight)^2)
+  return(var.est.mGR)
+}
+
+
+var.mh.Sato <- function(n11k, n10k, n.1k, n.0k){
+  weight <- n.1k * n.0k/ (n.1k + n.0k)
+  Pk = (n.1k^2 * n10k - n.0k^2*n11k +n.1k*n.0k * (n.0k-n.1k)/2)/ (n.1k + n.0k)^2
+  Qk = (n11k * (n.0k-n10k)+n10k *(n.1k - n11k))/2/(n.1k  + n.0k)
+  delta <- n11k / n.1k - n10k / n.0k
+  delta_all <- weighted.mean(delta, weight)
+  var.est.Sato = (delta_all*sum(Pk)+sum(Qk))/(sum(weight)^2)
+  return(var.est.Sato)
+}
+
+var.ate.nu = function(df, ATE.est){
+  n.1k = df$n.1k
+  n.0k = df$n.0k
+  n11k = df$n11k
+  n10k = df$n10k
+  y1.k = df$y1.k
+  y0.k = df$y0.k
+  y1.var.k = df$y1.var.k
+  y0.var.k = df$y0.var.k
+  
+  K = length(n11k)
+  n.k = n.1k + n.0k
+  n = sum(n.k)
+  pi1 = sum(n.1k) / n
+  pi0 = sum(n.0k) / n
+  rho.hat = n.k/n
+  delta.k = n11k / n.1k - n10k / n.0k
+  weights = n.1k*n.0k / (n.1k + n.0k)
+  
+  p1k = n11k / n.1k
+  p0k = n10k / n.0k
+  
+  p1k.sq = p1k^2 - ifelse(n.1k>1,y1.var.k,0) / n.1k
+  p0k.sq = p0k^2 - ifelse(n.0k>1,y0.var.k,0) / n.0k
+  
+  delta.k.sq = p1k.sq - 2*p1k * p0k + p0k.sq
+  tmp1<-rho.hat*(delta.k.sq-ATE.est^2)
+  tmp2<-(delta.k.sq - 2*delta.k*ATE.est +ATE.est^2)*pi1*pi0*(n.k-1)/n.k*(n.k-1-(4*n.k-6)*pi1*pi0)/n
+  var.est.nu = sum(tmp1[weights!=0])*pi1^2*pi0^2 + sum(tmp2[weights!=0])
+  
+  var.est.nu = var.est.nu / n / (sum(weights)/n)^2
+  
+  return(var.est.nu)
+}
+
+est.var.mh <- function(df.mh, estimand, ci_type){
+  mh_est <- estimate.mh(df.mh, estimand)
+  var_est <- with(df.mh,
+                  switch(ci_type,
+                         "GR"    = var.mh.GR(n11k, n10k, n.1k, n.0k),
+                         "mGR"   = var.mh.mGR(n11k, n10k, n.1k, n.0k) + (estimand=="ATE") * var.ate.nu(df.mh, mh_est),
+                         "Sato"  = var.mh.Sato(n11k, n10k, n.1k, n.0k),
+                         stop("Invalid ci_type. Choose from 'GR', 'mGR', or 'Sato'."))
+  )
+  return(list(estimate = mh_est, var_est = var_est))
+}
+
+# Estimate the MH risk difference or ATE
+# contrast is incorporated in adjust.MHModel as we don't have estimate for each treatment.
+#' @importFrom utils combn
+#' @exportS3Method
+adjust.MHModel <- function(model, data, ...){
+  
+  estimand <- model$estimand
+  ci_type <- model$ci_type
+  
+  treat_col <- data$treat_col
+  response_col <- data$response_col
+  strata <- data$joint_strata
+  
+  treat_levels <- data$treat_levels
+  n_treat <- length(treat_levels)
+  if(n_treat>2) {stop("The treatment variable must be binary.")}
+  if(estimand=="MH" & !all(data$response %in% c(0,1))) {
```

Comment:
I would prefer if these checks and the `stop()` statements were done in the `validate(data)` call using your `validate.RoboDataMH` method. For example, see this line https://github.com/mbannick/RobinCar/blob/main/R/Data.R#L64, where the `validate` method uses an additional argument (in your case would be `estimand`).